Improve return types for `render`, `renderAsync`, `renderFile` and `renderFileAsync`

nhaef opened the pull request:
Hey, I really like eta and I have been using it for a few projects already. I would love to contribute to this project. ❤

As already explained in [this issue](https://github.com/eta-dev/eta/issues/189), the types of the render functions can be improved by overloading them. There are essentially three branches and every branch defines their respective return type:

```ts
export default function render(
    template: string | TemplateFunction,
    data: object,
    config?: PartialConfig,
    cb?: CallbackFn
): string | Promise<string> | void {
    if (options.async) {
        if (cb) {
            // user provides callback function
            // return type is void
        } else {
            // user sets async option
            // return type is Promise<string>
        }
    } else {
        // return type is string
    }
}
```

The idea would be to define one function overload per return type so that the function can be consumed easily, without the need of any type assertion. The implementation signature is provided as well in order to prevent breaking changes.

I appreciate any feedback, ideas and suggestions.

This PR resolves https://github.com/eta-dev/eta/issues/189.

nebrelbug commented:
@nhaef thanks for working on this! I think this change will be really helpful. Is it ready for review yet?

nhaef commented:
Hi @nebrelbug, thanks for the feedback. It isn't ready for review yet, but it is almost there :-) I will update this PR later today

nhaef commented:
Okay, function overloads for `renderFile` and `renderFileAsync` have been added as well. The return type of both functions depends on the return type of `tryHandleCache`. This method is pretty much structured like this:

```ts
function tryHandleCache(data: object, options: EtaConfigWithFilename, cb: CallbackFn | undefined) {
    if (cb) {
        // return type is void
    } else {
        // return type is Promise<string>
    }
}
```

The function overloads of `renderFile` and `renderFileAsync` are specified so that the return type is `void` if the user provides a CallbackFn, otherwise the return type is `Promise<string>`. Similar to the previous changes, I did not remove the existing function overloads with the optional CallbackFn so that this PR does not introduce any breaking changes.

nhaef commented:
Here is a preview of the inferred types.

<details>
  <summary>click me</summary>

  ### Eta.render
  ```ts
  Eta.render(templateFn, data); // string
  Eta.render(templateFn, data, config); // string
  Eta.render(templateFn, data, { ...config, async: true }); // Promise<string>
  Eta.render(templateFn, data, { ...config, async: true }, callbackFn); // void
  ```
  
  ### Eta.renderAsync
  ```ts
  Eta.renderAsync(templateFn, data); // Promise<string>
  Eta.renderAsync(templateFn, data, config); // Promise<string>
  Eta.renderAsync(templateFn, data, { ...config, async: true }); // Promise<string>
  Eta.renderAsync(templateFn, data, { ...config, async: true }, callbackFn); // void
  ```
  
  ### Eta.renderFile
  ```ts
  Eta.renderFile(fileName, data, config); // Promise<string>
  Eta.renderFile(fileName, data, config, callbackFn); // void
  Eta.renderFile(fileName, { ...data, ...config }); // Promise<string>
  Eta.renderFile(fileName, { ...data, ...config }, callbackFn); // void
  ```
  
  ### Eta.renderFileAsync
  ```ts
  Eta.renderFileAsync(fileName, data, config); // Promise<string>
  Eta.renderFileAsync(fileName, data, config, callbackFn); // void
  Eta.renderFileAsync(fileName, { ...data, ...config }); // Promise<string>
  Eta.renderFileAsync(fileName, { ...data, ...config }, callbackFn); // void
  ```
</details>

nhaef commented:
The changes are now ready for review. I would especially be interested if I should add JSDoc annotations for each function overload or not. Looking forward to your feedback 🙂

nebrelbug commented on src/file-handlers.ts:
```diff
@@ -170,13 +170,28 @@ function includeFile(path: string, options: EtaConfig): [TemplateFunction, EtaCo
 
+function renderFile(filename: string, data: DataObj, cb: CallbackFn): void
+
 function renderFile(filename: string, data: DataObj, cb?: CallbackFn): Promise<string> | void
```
Could we remove this line?

nhaef replied:
yes, should be resolved, see https://github.com/eta-dev/eta/pull/199/commits/fcc1845dc1da2d105283e50aa44f60350ec010d6

nebrelbug commented on src/file-handlers.ts:
```diff
@@ -268,13 +283,28 @@ function renderFile(
 
+function renderFileAsync(filename: string, data: DataObj, cb: CallbackFn): void
+
 function renderFileAsync(filename: string, data: DataObj, cb?: CallbackFn): Promise<string> | void
```
Could we remove this line?

nhaef replied:
yes, should be resolved, see https://github.com/eta-dev/eta/pull/199/commits/fcc1845dc1da2d105283e50aa44f60350ec010d6

nebrelbug commented:
@nhaef thanks so much for your work here! Sorry I've just gotten to reviewing this now. As I commented above, it seems to me that we should be able to remove some old function overloads. Is that correct?

nebrelbug commented on src/file-handlers.ts:
```diff
@@ -268,13 +283,28 @@ function renderFile(
+function renderFileAsync(
+  filename: string,
+  data: DataObj,
+  config: PartialConfig,
```
I wonder if we should change this to `config?:`, and then delete lines 299-304

nhaef replied:
We can't mark the parameter as optional using `?` here, since it is followed by another, in that case, required parameter, which is `cb`. What would be possible is to allow undefined (`PartialConfig | undefined`). The types would still be correctly inferred in each [scenario mentioned above](https://github.com/eta-dev/eta/pull/199#issuecomment-1353727149).

However, I am not too sure about removing line 299-304, mainly because I think this introduces a breaking change for certain use cases. To give a theoretical example:

```ts
const optionalCallbackFn: CallbackFn | undefined = ...;
Eta.renderFileAsync(fileName, { ...data, ...config }, callbackFn);
```

Would not compile if we decide to remove 299-304. If we keep it, it would compile and the type would be inferred as `Promise<string> | void`.

What do you think?

nebrelbug commented on src/render.ts:
```diff
@@ -45,6 +45,83 @@ function handleCache(template: string | TemplateFunction, options: EtaConfig): T
+    data: object,
+    config?: PartialConfig,
+    cb?: CallbackFn
+): string | Promise<string> | void
```
I think we can delete lines 119-124.

nhaef replied:
Removing the implementation signature from the function overloads would introduce a breaking change, for example the following code will not compile anymore (similar to above):

```ts
const optionalCallbackFn: CallbackFn | undefined = ...;
Eta.render(templateFn, data, { ...config, async: true }, optionalCallbackFn);
```

That is why I initially kept it, but if you want I can remove it of course.

nebrelbug commented on src/render.ts:
```diff
@@ -98,6 +215,12 @@ export default function render(
+    data: object,
+    config?: PartialConfig,
+    cb?: CallbackFn
+): string | Promise<string> | void
```
Think we should be able to delete lines 218-223, since those definitions are stated in lines 225-230.

nhaef replied:
Oh, I think there might be a misunderstanding. Line 225-230 is the implementation signature, thus it will not be visible by our users.

> The signature of the implementation is not visible from the outside. [..]

See https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

If we want to make the implementation signature visible (for backwards compatibility in our case), we basically need to repeat it so that typescript knows we intend to make it visible to our users

nebrelbug replied:
> The signature of the implementation is not visible from the outside.

@nhaef you're right, I didn't realize that Typescript had this behavior! You learn something new every day, I guess :)

nhaef commented:
> @nhaef thanks so much for your work here! Sorry I've just gotten to reviewing this now. As I commented above, it seems to me that we should be able to remove some redundant function overloads. Is that correct?

No problem and thanks for the review :) And also sorry from my side as well, I was busy the last few days.

I removed some unnecessary function overloads and updated the conversations. I think some of the conversations which are still open can be summed up with the question "Should we make the implementation signature visible for backwards compatibility?".

nebrelbug commented:
@nhaef I actually didn't realize that implementation signatures were not visible to users. With that knowledge, this PR looks great. I'll merge right now!

nebrelbug commented:
@nhaef do you mind if I add you to the contributors section in the README?

nhaef commented:
> @nhaef do you mind if I add you to the contributors section in the README?

not at all, I'd be honored :D

nebrelbug commented:
@all-contributors please add @nhaef for code
